File: example/unity_client/Assets/Scripts/Handler/Dungeon/Handler_CreateDungeon.cs
using UnityEngine;
using System.Collections.Generic;

namespace Handler
{
	namespace Dungeon
	{
		class Handler_CreateDungeon
		{
			public static void SendMsg()
			{
				GameManager.Instance.session.RegisterHandler<Message.Dungeon.MsgSvrCli_CreateDungeon_Ans>(Handler_CreateDungeon.OnRecv);

				Message.Dungeon.MsgCliSvr_CreateDungeon_Req req = new Message.Dungeon.MsgCliSvr_CreateDungeon_Req();
				GameManager.Instance.session.SendMsg(req, true);
			}

			public static void OnRecv(Message.Dungeon.MsgSvrCli_CreateDungeon_Ans ans)
			{
				GameManager.Instance.session.UnregisterHandler<Message.Dungeon.MsgSvrCli_CreateDungeon_Ans>(Handler_CreateDungeon.OnRecv);
				GameObject go = new GameObject();
				go.name = "Dungeon";
				go.transform.SetParent(GameManager.Instance.scenes.dungeon_main.transform, false);

				Component.Dungeon dungeon = go.AddComponent<Component.Dungeon>();
				dungeon.width = ans.width;
				dungeon.height = ans.height;
				dungeon.tiles = new List<Tile>(new Tile[dungeon.width * dungeon.height]);

				for (int y = 0; y < dungeon.height; y++)
				{
					for (int x = 0; x < dungeon.width; x++)
					{
						if (Message.DungeonTileType.Invalid == ans.tiles[y * dungeon.width + x])
						{
							continue;
						}

						Tile.Data tileData = new Tile.Data();
						tileData.index = y * dungeon.width + x;
						tileData.type = ans.tiles[tileData.index];
						tileData.position = new Vector2Int(x, y);
						
						GameObject obj = new GameObject();
						obj.transform.SetParent(dungeon.transform, false);
						obj.transform.localPosition = new Vector3(x, y, 0.0f);
						//obj.SetActive(false);

						Tile tile = obj.AddComponent<Tile>();
						tile.Init(tileData);
						dungeon.tiles[y * dungeon.width + x] = tile;
					}
				}

				GameManager.Instance.scenes.dungeon_main.dungeon = dungeon;
				Unit player = GameManager.Instance.scenes.dungeon_main.player;
				player.position = new Vector2Int(ans.start.x, ans.start.y);

				foreach (var monster in ans.monsters)
				{
					GameObject obj = new GameObject();
					obj.name = "Monster_" + monster.seq;
					Unit unit = obj.AddComponent<Unit>();
					unit.seq = monster.seq;
					unit.position = new Vector2Int(monster.position.x, monster.position.y);
					obj.transform.SetParent(GameManager.Instance.scenes.dungeon_main.transform);
					obj.AddComponent<Monster>();
					GameManager.Instance.scenes.dungeon_main.monsters[unit.seq] = unit;
				}
			}
		}
	}
}

File: example/LobbyServer/Handler/Dungeon/Handler_CreateDungeon.cpp
#include "Handler_CreateDungeon.h"
#include "../../Component/UserData.h"
#include "../../Component/Dungeon/Dungeon.h"
#include "../../Component/Dungeon/Unit.h"
#include "../../Component/Dungeon/Unit/Monster.h"
#include <Gamnet/Library/Random.h>

namespace Handler { namespace Dungeon {

Handler_CreateDungeon::Handler_CreateDungeon()
{
}

Handler_CreateDungeon::~Handler_CreateDungeon()
{
}

void Handler_CreateDungeon::Recv_Req(const std::shared_ptr<UserSession>& session, const Message::Dungeon::MsgCliSvr_CreateDungeon_Req& req)
{
	Message::Dungeon::MsgSvrCli_CreateDungeon_Ans ans;
	ans.error_code = Message::ErrorCode::Success;
	
	try {
		LOG(DEV, "Message::Dungeon::MsgCliSvr_CreateDungeon_Req()");
		if(nullptr == session->GetComponent<Component::UserData>())
		{
			throw GAMNET_EXCEPTION(Message::ErrorCode::InvalidUserError);
		}

		std::shared_ptr<Component::Dungeon::Meta> meta = Gamnet::Singleton<Component::Dungeon::Manager>::GetInstance().FindMeta(1);
		std::shared_ptr<Component::Dungeon::Data> dungeon = meta->CreateInstance();
		session->AddComponent<Component::Dungeon::Data>(dungeon);
		session->strand = dungeon->strand;
		
		Vector2Int start = dungeon->start->rect.Center();
		Vector2Int end = dungeon->end->rect.Center();
		std::shared_ptr<Unit> player = std::make_shared<Unit>(dungeon);
		player->AddComponent<Player>();
		player->SetPosition(start);
		dungeon->player = player;
		/*
		std::shared_ptr<Unit> monster = Gamnet::Singleton<Component::Monster::Manager>::GetInstance().CreateInstance(1);
		monster->dungeon = dungeon;
		monster->SetPosition(Vector2Int((int)end.x, (int)end.y));
		dungeon->monster.insert(std::make_pair(monster->seq, monster));
		*/
		ans.width = dungeon->GetRect().width;
		ans.height = dungeon->GetRect().height;
		ans.start.x = dungeon->player->position.x;
		ans.start.y = dungeon->player->position.y;
		
		for(auto tile : dungeon->GetAllTiles())
		{
			ans.tiles.push_back(tile->type);
		}

		for(auto itr : dungeon->monster)
		{
			std::shared_ptr<Unit> monster = itr.second;
			Message::Monster msgMonster;
			msgMonster.seq = monster->seq;
			msgMonster.position.x = monster->position.x;
			msgMonster.position.y = monster->position.y;
			ans.monsters.push_back(msgMonster);
		}
	}
	catch (const Gamnet::Exception& e)
	{
		LOG(Gamnet::Log::Logger::LOG_LEVEL_ERR, e.what());
		ans.error_code = (Message::ErrorCode)e.error_code();
	}
	LOG(DEV, "Message::Dungeon::MsgSvrCli_CreateDungeon_Ans(error_code:", (int)ans.error_code, ")");
	Gamnet::Network::Tcp::SendMsg(session, ans);
}

GAMNET_BIND_TCP_HANDLER(
	UserSession,
	Message::Dungeon::MsgCliSvr_CreateDungeon_Req,
	Handler_CreateDungeon, Recv_Req,
	HandlerStatic
);

void Test_CreateDungeon_Req(const std::shared_ptr<TestSession>& session)
{
	Message::Dungeon::MsgCliSvr_CreateDungeon_Req req;
	Gamnet::Test::SendMsg(session, req);
}

void Test_CreateDungeon_Ans(const std::shared_ptr<TestSession>& session, const std::shared_ptr<Gamnet::Network::Tcp::Packet>& packet)
{
	Message::Dungeon::MsgSvrCli_CreateDungeon_Ans ans;
	try {
		if (false == Gamnet::Network::Tcp::Packet::Load(ans, packet))
		{
			throw GAMNET_EXCEPTION(Message::ErrorCode::MessageFormatError, "message load fail");
		}
		//		LOG(INF, "[", session->link->link_manager->name, "/", session->link->link_key, "/", session->session_key, "] Test_UserUser_CreateDungeonLobby_Ans");
	}
	catch (const Gamnet::Exception& e) {
		LOG(Gamnet::Log::Logger::LOG_LEVEL_ERR, e.what());
	}

	for(int y=0; y<ans.height; y++)
	{
		for(int x=0; x<ans.width; x++)
		{
			switch(ans.tiles[y * ans.width + x])
			{
			case Tile::Type::Door :
				std::cout << "+";
				break;
			case Tile::Type::StairDown :
				std::cout << "G";
				break;
			case Tile::Type::Wall :
				std::cout << "#";
				break;
			case Tile::Type::Floor:
				std::cout << ".";
				break;
			default :
				std::cout << " ";
				break;
			}
		}
		std::cout << std::endl;
	}

	session->Next();
}

GAMNET_BIND_TEST_HANDLER(
	TestSession, "Test_Dungeon_CreateDungeon",
	Message::Dungeon::MsgCliSvr_CreateDungeon_Req, Test_CreateDungeon_Req,
	Message::Dungeon::MsgSvrCli_CreateDungeon_Ans, Test_CreateDungeon_Ans
);

}}